urlac47192b66_ggrandes_bouncer_tgz-pJ8-org_javastack_bouncer_ClusterTestJ8/src/org/javastack/bouncer/PlainServer.java

```diff
@@ -0,0 +1,252 @@
+		volatile boolean shutdown = false;
+
+		@Override
+		public void setShutdown() {
+			shutdown = true;
+			context.closeSilent(listen);
+		}
+
+		@Override
+		public void run() {
+			try {
+				listen = inboundAddress.listen();
+				Log.info(this.getClass().getSimpleName() + " started: " + inboundAddress);
+				while (!shutdown) {
+					try {
+						final Socket client = listen.accept();
+						try {
+							context.registerSocket(client);
+							final Integer pReadTimeout = inboundAddress.getOpts().getInteger(
+									Options.P_READ_TIMEOUT);
+							if (pReadTimeout != null) {
+								client.setSoTimeout(pReadTimeout);
+							}
+							if (client instanceof SSLSocket) {
+								((SSLSocket) client).startHandshake();
+							}
+							Log.info(this.getClass().getSimpleName() + " New client from=" + client + " "
+									+ SSLFactory.getSocketProtocol(client));
+							context.submitTask(
+									new PlainConnector(client, inboundAddress.getOpts()),
+									"ForwardConnect[" + inboundAddress + "|"
+											+ IOHelper.socketRemoteToString(client) + "]",
+									(((long) client.getPort() << 48) | ClientId.newId()));
+						} catch (Exception e) {
+							Log.error(this.getClass().getSimpleName() + " Exception: " + e.toString(), e);
+							context.closeSilent(client);
+						}
+					} catch (SocketTimeoutException e) {
+						continue;
+					} catch (Exception e) {
+						if (!listen.isClosed()) {
+							Log.error(this.getClass().getSimpleName() + " " + e.toString(), e);
+						}
+					}
+				}
+			} catch (IOException e) {
+				Log.error(this.getClass().getSimpleName() + " " + e.toString());
+			} catch (Exception e) {
+				Log.error(this.getClass().getSimpleName() + " Generic exception", e);
+			} finally {
+				Log.info(this.getClass().getSimpleName() + " await end");
+				context.awaitShutdown(this);
+				Log.info(this.getClass().getSimpleName() + " end");
+			}
+		}
+	}
+
+	class PlainConnector implements Shutdownable, Runnable {
+		final Socket client;
+		final Options options;
+		Socket remote = null;
+		volatile boolean shutdown = false;
+
+		PlainConnector(final Socket client, final Options options) {
+			this.client = client;
+			this.options = options;
+		}
+
+		@Override
+		public void setShutdown() {
+			shutdown = true;
+			close();
+		}
+
+		void close() {
+			// FIXME: java.lang.NullPointerException: SocketRegistrator.unregisterSocket(Socket==null)
+			if (client != null)
+				context.closeSilent(client);
+			if (remote != null)
+				context.closeSilent(remote);
+		}
+
+		@Override
+		public void run() {
+			Log.info(this.getClass().getSimpleName() + " started: " + outboundAddress);
+			try {
+				try {
+					context.getStatistics().incTryingConnections();
+					remote = outboundAddress.connectFrom(client.getInetAddress());
+				} finally {
+					context.getStatistics().decTryingConnections();
+				}
+				if (remote == null)
+					throw new ConnectException("Unable to connect to " + outboundAddress);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Exception [ConnectException] is explicitly thrown on this line but is caught in the outer catch clause with generic exception on line [149].  Consider adding a catch clause of this specific exception in the outer try statement.